systems/ros/ros_publisher_system.h

```diff
@@ -0,0 +1,198 @@
+    std::unordered_set<drake::systems::TriggerType, drake::DefaultHash>;
+
+/**
+ * Publishes an ROS message containing information from its input port.
+ *
+ * @ingroup message_passing
+ */
+template <typename RosMessage>
+class RosPublisherSystem : public drake::systems::LeafSystem<double> {
+ public:
+  DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(RosPublisherSystem)
+
+  /**
+   * A factory method that returns an %RosPublisherSystem that takes
+   * Value<RosMessage> message objects on its sole abstract-valued input port.
+   *
+   * @param[in] topic The ROS topic on which to publish.
+   *
+   * @param node_handle The ROS context.
+   *
+   * @param publish_triggers Set of triggers that determine when messages will
+   * be published. Supported TriggerTypes are {kForced, kPeriodic, kPerStep}.
+   * Will throw an error if empty or if unsupported types are provided.
+   *
+   * @param publish_period Period that messages will be published (optional).
+   * publish_period should only be non-zero if one of the publish_triggers is
+   * kPeriodic.
+   */
+  static std::unique_ptr<RosPublisherSystem<RosMessage>> Make(
+      const std::string& topic, ros::NodeHandle* node_handle,
+      double publish_period = 0.0) {
+    return std::make_unique<RosPublisherSystem<RosMessage>>(topic, node_handle,
+      publish_period);
+  }
+
+  /**
+   * A factory method that returns an %RosPublisherSystem that takes
+   * Value<RosMessage> message objects on its sole abstract-valued input port.
+   *
+   * @param[in] topic The ROS topic on which to publish.
+   *
+   * @param node_handle The ROS context.
+   *
+   * @param publish_period Period that messages will be published (optional).
+   * If the publish period is zero, RosPublisherSystem will use per-step
+   * publishing instead; see LeafSystem::DeclarePerStepPublishEvent().
+   */
+  static std::unique_ptr<RosPublisherSystem<RosMessage>> Make(
+      const std::string& topic, ros::NodeHandle* node_handle,
+      const TriggerTypeSet& publish_triggers, double publish_period = 0.0) {
+    return std::make_unique<RosPublisherSystem<RosMessage>>(topic, node_handle,
+      publish_triggers, publish_period);
+  }
+
+  /**
+   * A constructor for an %RosPublisherSystem that takes ROS message objects on
+   * its sole abstract-valued input port. The ROS message type is determined by
+   * the template type.
+   *
+   * @param[in] topic The ROS topic on which to publish.
+   *
+   * @param node_handle The ROS context.
+   *
+   * @param publish_triggers Set of triggers that determine when messages will
+   * be published. Supported TriggerTypes are {kForced, kPeriodic, kPerStep}.
+   * Will throw an error if empty or if unsupported types are provided.
+   *
+   * @param publish_period Period that messages will be published (optional).
+   * publish_period should only be non-zero if one of the publish_triggers is
+   * kPeriodic.
+   */
+  RosPublisherSystem(const std::string& topic, ros::NodeHandle* node_handle,
```

Comment:
Any particular reason for implementing a large function in the header file?

Reply:
It's implemented in a header because it's templated on message type, but `ros_publisher_system` does not know, at compile time, what explicit types to compile (only those that link against it do).